File: studio/security.py
"""Password hashing (stdlib PBKDF2) and HS256 JWT issue/verify."""

import base64
import hashlib
import hmac
import os
import secrets
import time

import jwt as pyjwt

_ITERATIONS = 600_000  # OWASP-recommended floor for PBKDF2-SHA256
_ISSUER = "dharwin-auth"
_AUDIENCE = "dharwin-api"
_LEEWAY_S = 30
TOKEN_TTL_S = 24 * 3600


class TokenError(Exception):
    pass


def _secret():
    value = os.environ.get("AUTH_JWT_SECRET", "").strip()
    if not value:
        raise RuntimeError("AUTH_JWT_SECRET is not set")
    return value


def hash_password(password, salt=None):
    salt = salt or secrets.token_hex(16)
    digest = hashlib.pbkdf2_hmac(
        "sha256", password.encode("utf-8"), salt.encode("utf-8"), _ITERATIONS
    )
    return base64.b64encode(digest).decode("ascii"), salt


def verify_password(password, password_hash, salt):
    candidate, _ = hash_password(password, salt)
    return hmac.compare_digest(candidate, password_hash)


def issue_jwt(user_id, now=None):
    issued = int(now if now is not None else time.time())
    return pyjwt.encode(
        {
            "sub": user_id,
            "iat": issued,
            "exp": issued + TOKEN_TTL_S,
            "iss": _ISSUER,
            "aud": _AUDIENCE,
        },
        _secret(),
        algorithm="HS256",
    )


def verify_jwt(token):
    try:
        payload = pyjwt.decode(
            token,
            _secret(),
            algorithms=["HS256"],
            issuer=_ISSUER,
            audience=_AUDIENCE,
            leeway=_LEEWAY_S,
            options={"require": ["sub", "exp", "iss", "aud"]},
        )
    except pyjwt.PyJWTError as exc:
        raise TokenError(str(exc)) from exc
    return payload["sub"]

File: src/server/security.ts
import { jwtVerify } from "jose";

const ISSUER = "dharwin-auth";
const AUDIENCE = "dharwin-api";

export class TokenError extends Error {}

function secret(): Uint8Array {
  const value = process.env.AUTH_JWT_SECRET?.trim();
  if (!value) throw new Error("AUTH_JWT_SECRET is not set");
  return new TextEncoder().encode(value);
}

/** JWT verify for proxy (Node.js runtime in Next.js 16). */
export async function verifyJwt(token: string): Promise<string> {
  try {
    const { payload } = await jwtVerify(token, secret(), {
      issuer: ISSUER,
      audience: AUDIENCE,
      clockTolerance: 30,
    });
    const sub = payload.sub;
    if (typeof sub !== "string" || !sub) throw new TokenError("invalid token subject");
    return sub;
  } catch {
    throw new TokenError("invalid or expired token");
  }
}
